### PR TITLE
ENT-3713: [PART 1/2] Fix duplicate snapshot generation when rolling hourly snapshots

### PR DESCRIPTION
**Pull request by mstead**
[BUG ENT-3713](https://issues.redhat.com/browse/ENT-3713)

**NOTE: THERE WILL BE A FOLLOW ON PR THAT WILL ADD CONSTRAINTS TO THE tally_snapshots TABLE AND TO CLEAN UP EXISTING DUPLICATE DATA. THIS PR WILL AT LEAST STOP THE BLEEDING.**

Roll hourly snapshots based on date range of calculations

Duplicate snapshots were getting created when events were processed
and snapshots were rolled due to date ranges.

Given that the date range can be modified when the calculations are
being done, use the range provided by the resulting calculations so
we are sure that the entire range is covered.

## Test
Start with a fresh database and ensure you have prometheus server available.

```bash
$ dropdb -U rhsm-subscriptions rhsm-subscriptions && createdb -U rhsm-subscriptions rhsm-subscriptions
```

**From the 'develop' branch**
Deploy the app
```bash
$ OPENSHIFT_BILLING_MODEL_FILTER='' DEV_MODE=true PROM_URL="http://localhost:8082/api/v1" ./gradlew clean bootRun
```

Pull in some events from prometheus metrics
```bash
$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=openshiftJmxBean,type=OpenShiftJmxBean","operation":"performCustomOpenshiftMeteringForAccount(java.lang.String, java.lang.String, java.lang.Integer)","arguments":["6661312", null, 50000]}'
```

Generate the snapshots for the account a few times.
```bash
$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["6661312", "2021-03-30T13:00:00Z", "2021-04-01T13:00:00Z"]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["6661312", "2021-03-30T13:00:00Z", "2021-04-01T13:00:00Z"]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["6661312", "2021-03-30T13:00:00Z", "2021-04-01T13:00:00Z"]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["6661312", "2021-03-30T13:00:00Z", "2021-04-01T13:00:00Z"]}'
```

Check the database to see the duplicate snapshots. Each hour will have an entry for each time you ran the tally.
```sql
rhsm-subscriptions=# select s.id, s.account_number, s.snapshot_date, s.usage, s.sla, s.granularity, m.* from tally_snapshots s, tally_measurements m where s.id = m.snapshot_id and s.account_number = '6661312' and s.granularity='HOURLY' and EXTRACT(MONTH from s.snapshot_date) = 3 and s.sla='_ANY' and s.usage='_ANY' and s.product_id='OpenShift-metrics' and measurement_type='TOTAL' order by snapshot_date;
                  id                  | account_number |     snapshot_date      | usage | sla  | granularity |             snapshot_id              | measurement_type |  uom  |       value        
--------------------------------------+----------------+------------------------+-------+------+-------------+--------------------------------------+------------------+-------+--------------------
 414111d9-deb1-41e7-831f-95b7ca74ae8e | 6661312        | 2021-03-21 21:00:00-03 | _ANY  | _ANY | HOURLY      | 414111d9-deb1-41e7-831f-95b7ca74ae8e | TOTAL            | CORES | 33.230769230769226
 84cddac8-8353-4a39-92a8-aed677ce6360 | 6661312        | 2021-03-21 21:00:00-03 | _ANY  | _ANY | HOURLY      | 84cddac8-8353-4a39-92a8-aed677ce6360 | TOTAL            | CORES | 33.230769230769226
 06b367d1-1f01-45a2-a7ac-671de651ba9e | 6661312        | 2021-03-21 21:00:00-03 | _ANY  | _ANY | HOURLY      | 06b367d1-1f01-45a2-a7ac-671de651ba9e | TOTAL            | CORES | 33.230769230769226
 b91d1a26-256d-4163-8408-3ea0c60616a4 | 6661312        | 2021-03-21 21:00:00-03 | _ANY  | _ANY | HOURLY      | b91d1a26-256d-4163-8408-3ea0c60616a4 | TOTAL            | CORES | 33.230769230769226
 89d6ee6c-dff6-4330-aa58-3c50ff0e02db | 6661312        | 2021-03-21 22:00:00-03 | _ANY  | _ANY | HOURLY      | 89d6ee6c-dff6-4330-aa58-3c50ff0e02db | TOTAL            | CORES |                 36
 515669a7-b9a5-48da-90c6-4eb424fc6056 | 6661312        | 2021-03-21 22:00:00-03 | _ANY  | _ANY | HOURLY      | 515669a7-b9a5-48da-90c6-4eb424fc6056 | TOTAL            | CORES |                 36
 36559287-0ff7-4969-954b-a3b2536f6b14 | 6661312        | 2021-03-21 22:00:00-03 | _ANY  | _ANY | HOURLY      | 36559287-0ff7-4969-954b-a3b2536f6b14 | TOTAL            | CORES |                 36
 28ded4c7-59fd-41aa-978c-076d2b52413d | 6661312        | 2021-03-21 22:00:00-03 | _ANY  | _ANY | HOURLY      | 28ded4c7-59fd-41aa-978c-076d2b52413d | TOTAL            | CORES |                 36
 4473e830-b26d-4045-965e-c53ff9bedb28 | 6661312        | 2021-03-21 23:00:00-03 | _ANY  | _ANY | HOURLY      | 4473e830-b26d-4045-965e-c53ff9bedb28 | TOTAL            | CORES |                 36
 ee18af66-4d5a-4cfe-9e5a-5dbf90f88c9b | 6661312        | 2021-03-21 23:00:00-03 | _ANY  | _ANY | HOURLY      | ee18af66-4d5a-4cfe-9e5a-5dbf90f88c9b | TOTAL            | CORES |                 36

-- snip --
```

Clean the database
```sql
TRUNCATE tally_snapshots CASCADE;
```

**DEPLOY THIS BRANCH**
Kick off the tally as you did above (multiple times) and verify in the database with the same query that there are no longer duplicates generated.

```sql
rhsm-subscriptions=# select s.id, s.account_number, s.snapshot_date, s.usage, s.sla, s.granularity, m.* from tally_snapshots s, tally_measurements m where s.id = m.snapshot_id and s.account_number = '6661312' and s.granularity='HOURLY' and EXTRACT(MONTH from s.snapshot_date) = 3 and s.sla='_ANY' and s.usage='_ANY' and s.product_id='OpenShift-metrics' and measurement_type='TOTAL' order by snapshot_date;
                  id                  | account_number |     snapshot_date      | usage | sla  | granularity |             snapshot_id              | measurement_type |  uom  |       value        
--------------------------------------+----------------+------------------------+-------+------+-------------+--------------------------------------+------------------+-------+--------------------
 884aa60c-7e08-4f79-8e0c-85ce978c7ddc | 6661312        | 2021-03-21 21:00:00-03 | _ANY  | _ANY | HOURLY      | 884aa60c-7e08-4f79-8e0c-85ce978c7ddc | TOTAL            | CORES | 33.230769230769226
 be0864b7-d6b3-4425-9bf7-e0e63fe89ec0 | 6661312        | 2021-03-21 22:00:00-03 | _ANY  | _ANY | HOURLY      | be0864b7-d6b3-4425-9bf7-e0e63fe89ec0 | TOTAL            | CORES |                 36
 f748956b-c48a-4a91-a49f-bea484208800 | 6661312        | 2021-03-21 23:00:00-03 | _ANY  | _ANY | HOURLY      | f748956b-c48a-4a91-a49f-bea484208800 | TOTAL            | CORES |                 36
 8fd8684d-8f13-4e70-a65c-adedcc558b33 | 6661312        | 2021-03-22 00:00:00-03 | _ANY  | _ANY | HOURLY      | 8fd8684d-8f13-4e70-a65c-adedcc558b33 | TOTAL            | CORES |                 36
 b3c61330-afe6-4a18-a7e7-bb179b5f2888 | 6661312        | 2021-03-22 01:00:00-03 | _ANY  | _ANY | HOURLY      | b3c61330-afe6-4a18-a7e7-bb179b5f2888 | TOTAL            | CORES |                 36
 dee55d40-8437-4d14-8b3e-5706432dc364 | 6661312        | 2021-03-22 02:00:00-03 | _ANY  | _ANY | HOURLY      | dee55d40-8437-4d14-8b3e-5706432dc364 | TOTAL            | CORES |                 36
 8eb1ffca-8c26-4e14-bb7f-7add73f5213c | 6661312        | 2021-03-22 03:00:00-03 | _ANY  | _ANY | HOURLY      | 8eb1ffca-8c26-4e14-bb7f-7add73f5213c | TOTAL            | CORES |                 36

-- snip --
```

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/util/DateRange.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * A simple class that represents a date range.
+ */
+@Getter
+@Setter
+public class DateRange {
+
```
Comment (dmckoske) [inline src/main/java/org/candlepin/subscriptions/util/DateRange.java:36]: I like this.  We have needed to isolate date ranges for awhile.  Team, keep this in mind for whenever you touch files which use date ranges.

**Review by dmckoske — APPROVED** [review]
Looks good.  Minor comment for the team.

File: src/main/java/org/candlepin/subscriptions/util/DateRange.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    private final OffsetDateTime startDate;
+    private final OffsetDateTime endDate;
+
+    public DateRange(OffsetDateTime startDate, OffsetDateTime endDate) {
+        verifyRange(startDate, endDate);
+        this.startDate = startDate;
+        this.endDate = endDate;
+    }
+
+    public static DateRange from(Collection<OffsetDateTime> dates) {
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/util/DateRange.java:46]: I would like to see some Javadoc on this just so people don't have to run the logic in their heads.  Something like "Given a collection of dates, create a range spanning the earliest date in the collection to the latest date."

**Review by awood — APPROVED** [review]

**Comment by awood** [comment]
I had to run Sonar locally and it looks like it flagged a couple things on a test file.

Selection_010

File: src/main/java/org/candlepin/subscriptions/util/DateRange.java (hunk 2 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/util/DateRange.java:53]: DONE

File: src/main/java/org/candlepin/subscriptions/util/DateRange.java (hunk 1 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/util/DateRange.java:36]: :+1:

**Comment by mstead** [comment]
> I had to run Sonar locally and it looks like it flagged a couple things on a test file.
> 
> Selection_010

DONE

**Pull request merged**